qiskit/pulse/transforms/canonicalization.py

```diff
@@ -36,17 +36,36 @@ def block_to_schedule(block: ScheduleBlock) -> Schedule:
 
     Raises:
         UnassignedDurationError: When any instruction duration is not assigned.
+        PulseError: When the alignment context duration is shorter than the schedule duration.
+
+    .. note:: This transform may insert barriers in between contexts.
     """
     if not block.is_schedulable():
         raise UnassignedDurationError(
             'All instruction durations should be assigned before creating `Schedule`.'
             'Please check `.parameters` to find unassigned parameter objects.')
 
     schedule = Schedule(name=block.name, metadata=block.metadata)
+
     for op_data in block.instructions:
         if isinstance(op_data, ScheduleBlock):
             context_schedule = block_to_schedule(op_data)
+            if hasattr(op_data.alignment_context, 'duration'):
+                # context may have local scope duration, e.g. EquispacedAlignment for 1000 dt
+                post_buffer = op_data.alignment_context.duration - context_schedule.duration
+                if post_buffer < 0:
+                    raise PulseError(f'ScheduleBlock {op_data.name} has longer duration than '
+                                     'the specified context duration '
+                                     f'{context_schedule.duration} > {op_data.duration}.')
```

Comment:
I'm not sure if I'm understanding correctly. Here's my guess: if someone specifies that a length 100 schedule should go into a length 500 block, you'd add a barrier 400 steps past the end of that schedule. The placement of the 400dt worth of delay(s) would later be resolved by the alignment type. Is that right?

Reply:
Right. Strictly speaking isinstance check is not performed here. Instead, 500dt-length block is recognized as
```python
if hasattr(op_data.alignment_context, 'duration'):
```
https://github.com/nkanazawa1989/qiskit-terra/blob/53cd6bb2c7e6f2118543335091cdddbd4c491271/qiskit/pulse/transforms/canonicalization.py#L53

Note that padding with delay (placeholder) is permanently removed, because this just increases number of instructions and this may hurt performance as device scales. The barrier behaves as delays since block doesn't have insert. This will be removed at the final step of transformation. Thus the output schedule is much cleaner.